File: sdk/core/ios/Sources/Adapter/AdapterRegistry.swift
import Foundation

// MARK: - Adapter Protocol

/// Base adapter protocol
///
/// All ad network adapters must implement this protocol
public protocol AdNetworkAdapter {
    init()
    /// Adapter metadata
    var networkName: String { get }
    var version: String { get }
    var minSDKVersion: String { get }
    
    /// Initialize adapter with app-level configuration
    func initialize(config: [String: Any]) throws
    
    /// Load an ad
    func loadAd(
        placement: String,
        adType: AdType,
        config: [String: Any],
        completion: @escaping (Result<Ad, AdapterError>) -> Void
    )
    
    /// Check if adapter supports specific ad type
    func supportsAdType(_ adType: AdType) -> Bool
    
    /// Clean up adapter resources
    func destroy()
}

// MARK: - Adapter Registry

/// Adapter registry with dynamic loading
///
/// Features:
/// - Dynamic framework loading
/// - Version compatibility checking
/// - Lazy initialization
/// - Thread-safe adapter management
public final class AdapterRegistry {
    private var adapters: [String: AdapterWrapper] = [:]
    private var adapterClasses: [String: AdNetworkAdapter.Type] = [:]
    private let lock = NSLock()
    
    private let sdkVersion: String
    
    /// Adapter wrapper with metadata
    private struct AdapterWrapper {
        let adapter: AdNetworkAdapter
        var isInitialized: Bool
        var initError: String?
    }
    
    /// Initialize adapter registry
    public init(sdkVersion: String = "1.0.0") {
        self.sdkVersion = sdkVersion
        registerBuiltInAdapters()
    }
    
    // MARK: - Adapter Discovery
    
    /// Register built-in adapters that ship with the SDK target
    private func registerBuiltInAdapters() {
        let builtIns: [String: AdNetworkAdapter.Type] = [
            "admob": AdMobAdapter.self,
            "applovin": AppLovinAdapter.self,
            "unity": UnityAdsAdapter.self,
            "ironsource": IronSourceAdapter.self,
            "facebook": FacebookAdapter.self,
            "vungle": VungleAdapter.self,
            "chartboost": ChartboostAdapter.self,
            "pangle": PangleAdapter.self,
            "mintegral": MintegralAdapter.self,
            "adcolony": AdColonyAdapter.self,
            "tapjoy": TapjoyAdapter.self,
            "inmobi": InMobiAdapter.self,
            "fyber": FyberAdapter.self,
            "smaato": SmaatoAdapter.self,
            "amazon": AmazonAdapter.self
        ]
        adapterClasses.merge(builtIns) { _, new in new }
    }
    
    /// Register custom adapter
    public func registerAdapter(networkName: String, adapterClass: AdNetworkAdapter.Type) {
        lock.lock()
        adapterClasses[networkName] = adapterClass
        lock.unlock()
    }
    
    /// Section 3.1: Get count of registered adapters for debug panel
    public var registeredCount: Int {
        lock.lock()
        defer { lock.unlock() }
        return adapterClasses.count
    }
    
    /// Initialization diagnostics
    public struct InitializationStatus: Codable, Equatable {
        public let networkName: String
        public let registered: Bool
        public let initialized: Bool
        public let version: String?
        public let minSDKVersion: String?
        public let error: String?
    }
    
    /// Returns a list of initialization statuses for all known adapters
    public func getInitializationReport() -> [InitializationStatus] {
        lock.lock()
        defer { lock.unlock() }
        let known = Set(adapterClasses.keys)
        let union = Array(known.union(Set(adapters.keys))).sorted()
        return union.map { name in
            let wrapper = adapters[name]
            let cls = adapterClasses[name]
            let adapter: AdNetworkAdapter? = {
                if let w = wrapper { return w.adapter }
                if let c = cls { return c.init() }
                return nil
            }()
            return InitializationStatus(
                networkName: name,
                registered: cls != nil,
                initialized: wrapper?.isInitialized == true,
                version: adapter?.version,
                minSDKVersion: adapter?.minSDKVersion,
                error: wrapper?.initError
            )
        }
    }
    
    // MARK: - Adapter Management
    
    /// Get adapter instance (lazy initialization)
    public func getAdapter(networkName: String) -> AdNetworkAdapter? {
        lock.lock()
        defer { lock.unlock() }
        
        // Check if already initialized
        if let wrapper = adapters[networkName], wrapper.isInitialized {
            return wrapper.adapter
        }
        
        // Load adapter class
        guard let adapterClass = adapterClasses[networkName] else {
            return nil
        }
        
        // Create adapter instance
        let adapter = adapterClass.init()
        
        // Check version compatibility
        if !isCompatible(minVersion: adapter.minSDKVersion) {
            adapters[networkName] = AdapterWrapper(
                adapter: adapter,
                isInitialized: false,
                initError: "Incompatible SDK version"
            )
            return nil
        }
        
        adapters[networkName] = AdapterWrapper(
            adapter: adapter,
            isInitialized: false,
            initError: nil
        )
        
        return adapter
    }
    
    /// Initialize adapter
    public func initializeAdapter(networkName: String, config: [String: Any]) {
        lock.lock()
        defer { lock.unlock() }
        
        guard var wrapper = adapters[networkName] else { return }
        
        if wrapper.isInitialized {
            return
        }
        
        do {
            try wrapper.adapter.initialize(config: config)
            wrapper.isInitialized = true
            wrapper.initError = nil
            adapters[networkName] = wrapper
        } catch {
            wrapper.isInitialized = false
            wrapper.initError = error.localizedDescription
            adapters[networkName] = wrapper
        }
    }
    
    /// Check if adapter is initialized
    public func isInitialized(networkName: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        return adapters[networkName]?.isInitialized ?? false
    }
    
    /// Get all available adapters
    public func getAvailableAdapters() -> [String] {
        lock.lock()
        defer { lock.unlock() }
        return Array(adapterClasses.keys)
    }
    
    /// Get all initialized adapters
    public func getInitializedAdapters() -> [String] {
        lock.lock()
        defer { lock.unlock() }
        return adapters.filter { $0.value.isInitialized }.map { $0.key }
    }
    
    /// Get adapter initialization error
    public func getAdapterError(networkName: String) -> String? {
        lock.lock()
        defer { lock.unlock() }
        return adapters[networkName]?.initError
    }
    
    // MARK: - Version Compatibility
    
    /// Check version compatibility
    private func isCompatible(minVersion: String) -> Bool {
        return compareVersions(sdkVersion, minVersion) >= 0
    }
    
    /// Compare semantic versions
    private func compareVersions(_ version1: String, _ version2: String) -> Int {
        let parts1 = version1.split(separator: ".").compactMap { Int($0) }
        let parts2 = version2.split(separator: ".").compactMap { Int($0) }
        
        let maxLength = max(parts1.count, parts2.count)
        
        for i in 0..<maxLength {
            let part1 = i < parts1.count ? parts1[i] : 0
            let part2 = i < parts2.count ? parts2[i] : 0
            
            if part1 != part2 {
                return part1 < part2 ? -1 : 1
            }
        }
        
        return 0
    }
    
    /// Clean up all adapters
    public func destroy() {
        lock.lock()
        defer { lock.unlock() }
        
        for (_, wrapper) in adapters {
            wrapper.adapter.destroy()
        }
        
        adapters.removeAll()
        adapterClasses.removeAll()
    }
}

// MARK: - Adapter Error

/// Adapter errors
public enum AdapterError: Error, LocalizedError {
    case notInitialized
    case loadFailed(String)
    case unsupportedAdType
    case timeout
    case networkError
    
    public var errorDescription: String? {
        switch self {
        case .notInitialized:
            return "Adapter not initialized"
        case .loadFailed(let message):
            return "Ad load failed: \(message)"
        case .unsupportedAdType:
            return "Unsupported ad type"
        case .timeout:
            return "Ad load timeout"
        case .networkError:
            return "Network error"
        }
    }
}

// MARK: - Example Adapter Implementation

/// Example AdMob adapter implementation
public class AdMobAdapter: AdNetworkAdapter {
    public var networkName: String { "admob" }
    public var version: String { "1.0.0" }
    public var minSDKVersion: String { "1.0.0" }
    
    private var isInitialized = false
    
    public required init() {}
    
    public func initialize(config: [String: Any]) throws {
        guard config["app_id"] as? String != nil else {
            throw AdapterError.loadFailed("app_id required")
        }
        
        // Initialize AdMob SDK
        // GADMobileAds.sharedInstance().start(completionHandler: nil)
        
        isInitialized = true
    }
    
    public func loadAd(
        placement: String,
        adType: AdType,
        config: [String: Any],
        completion: @escaping (Result<Ad, AdapterError>) -> Void
    ) {
        guard isInitialized else {
            completion(.failure(.notInitialized))
            return
        }
        
        guard supportsAdType(adType) else {
            completion(.failure(.unsupportedAdType))
            return
        }
        
        // Load ad from AdMob
        // Implementation depends on AdMob SDK integration
        
        // Example failure
        completion(.failure(.loadFailed("Not implemented")))
    }
    
    public func supportsAdType(_ adType: AdType) -> Bool {
        switch adType {
        case .banner, .interstitial, .rewarded:
            return true
        default:
            return false
        }
    }
    
    public func destroy() {
        isInitialized = false
    }
}

// MARK: - Example AppLovin Adapter

/// Example AppLovin adapter implementation
public class AppLovinAdapter: AdNetworkAdapter {
    public var networkName: String { "applovin" }
    public var version: String { "1.0.0" }
    public var minSDKVersion: String { "1.0.0" }
    
    private var isInitialized = false
    
    public required init() {}
    
    public func initialize(config: [String: Any]) throws {
        guard config["sdk_key"] as? String != nil else {
            throw AdapterError.loadFailed("sdk_key required")
        }
        
        // Initialize AppLovin SDK
        // ALSdk.shared()?.initialize(withSdkKey: sdkKey)
        
        isInitialized = true
    }
    
    public func loadAd(
        placement: String,
        adType: AdType,
        config: [String: Any],
        completion: @escaping (Result<Ad, AdapterError>) -> Void
    ) {
        guard isInitialized else {
            completion(.failure(.notInitialized))
            return
        }
        
        guard supportsAdType(adType) else {
            completion(.failure(.unsupportedAdType))
            return
        }
        
        // Load ad from AppLovin
        completion(.failure(.loadFailed("Not implemented")))
    }
    
    public func supportsAdType(_ adType: AdType) -> Bool {
        switch adType {
        case .banner, .interstitial, .rewarded:
            return true
        default:
            return false
        }
    }
    
    public func destroy() {
        isInitialized = false
    }
}

// MARK: - Example Unity Ads Adapter

/// Example Unity Ads adapter implementation
public class UnityAdsAdapter: AdNetworkAdapter {
    public var networkName: String { "unity" }
    public var version: String { "1.0.0" }
    public var minSDKVersion: String { "1.0.0" }
    
    private var isInitialized = false
    
    public required init() {}
    
    public func initialize(config: [String : Any]) throws {
        guard let gameId = config["game_id"] as? String, !gameId.isEmpty else {
            throw AdapterError.loadFailed("game_id required")
        }
        // UnityAds.initialize(gameId)
        isInitialized = true
    }
    
    public func loadAd(
        placement: String,
        adType: AdType,
        config: [String : Any],
        completion: @escaping (Result<Ad, AdapterError>) -> Void
    ) {
        guard isInitialized else {
            completion(.failure(.notInitialized))
            return
        }
        guard supportsAdType(adType) else {
            completion(.failure(.unsupportedAdType))
            return
        }
        // Return a mock filled ad for testing purposes
        let creative: Creative = .banner(
            imageURL: "https://example.invalid/ad.png",
            clickURL: "https://example.invalid/click",
            width: 320,
            height: 50
        )
        let ad = Ad(
            adId: "mock-unity-\(placement)",
            placement: placement,
            adType: adType,
            creative: creative,
            networkName: networkName,
            cpm: 1.05,
            expiresAt: Date().addingTimeInterval(3600),
            metadata: [:]
        )
        completion(.success(ad))
    }
    
    public func supportsAdType(_ adType: AdType) -> Bool {
        switch adType {
        case .banner, .interstitial, .rewarded:
            return true
        default:
            return false
        }
    }
    
    public func destroy() {
        isInitialized = false
    }
}

File: sdk/core/android/src/main/kotlin/adapter/AdapterRegistry.kt
package com.rivalapexmediation.sdk.adapter

import android.content.Context
import com.rivalapexmediation.sdk.models.Ad
import com.rivalapexmediation.sdk.models.AdType
import com.rivalapexmediation.sdk.BuildConfig
import java.util.concurrent.ConcurrentHashMap

/**
 * Base adapter interface
 * 
 * All ad network adapters must implement this interface
 */
interface AdNetworkAdapter {
    /**
     * Adapter metadata
     */
    val networkName: String
    val version: String
    val minSDKVersion: String
    
    /**
     * Initialize adapter with app-level configuration
     */
    fun initialize(context: Context, config: Map<String, Any>)
    
    /**
     * Load an ad
     * 
     * @param placement Placement identifier
     * @param adType Type of ad to load
     * @param config Adapter-specific configuration
     * @param callback Callback for ad load result
     */
    fun loadAd(
        placement: String,
        adType: AdType,
        config: Map<String, Any>,
        callback: AdLoadCallback
    )
    
    /**
     * Check if adapter supports specific ad type
     */
    fun supportsAdType(adType: AdType): Boolean
    
    /**
     * Clean up adapter resources
     */
    fun destroy()
}

/**
 * Ad load callback
 */
interface AdLoadCallback {
    fun onAdLoaded(ad: Ad)
    fun onAdFailed(error: String, errorCode: Int)
}

/**
 * Adapter registry with dynamic loading
 * 
 * Features:
 * - Reflection-based adapter discovery
 * - Version compatibility checking
 * - Lazy initialization
 * - Thread-safe adapter management
 */
class AdapterRegistry(
    private val context: Context,
    private val sdkVersion: String
) {
    private val adapters = ConcurrentHashMap<String, AdapterWrapper>()
    private val adapterClasses = ConcurrentHashMap<String, Class<*>>()
    
    /**
     * Public diagnostics: how many adapters are registered (discovered or manually registered)
     */
    val registeredCount: Int get() = adapterClasses.size

    /** Return all registered adapter network names (sorted) */
    fun getAdapterNames(): List<String> = adapterClasses.keys().toList().sorted()
    
    /**
     * Adapter wrapper with metadata
     */
    private data class AdapterWrapper(
        val adapter: AdNetworkAdapter,
        val isInitialized: Boolean = false,
        val initError: String? = null
    )
    
    init {
        // Only attempt reflection-based discovery in debug or when explicitly enabled.
        if (isSandboxAdapterDiscoveryEnabled()) {
            discoverAdapters()
        }
    }
    
    /**
     * Discover available adapters via reflection
     */
    private fun discoverAdapters() {
        // Common adapter package names
        val adapterPackages = listOf(
            "com.rivalapexmediation.adapter.admob",
            "com.rivalapexmediation.adapter.applovin",
            "com.rivalapexmediation.adapter.ironsource",
            "com.rivalapexmediation.adapter.unity",
            "com.rivalapexmediation.adapter.facebook",
            "com.rivalapexmediation.adapter.vungle",
            "com.rivalapexmediation.adapter.chartboost",
            "com.rivalapexmediation.adapter.pangle",
            "com.rivalapexmediation.adapter.mintegral",
            "com.rivalapexmediation.adapter.adcolony",
            "com.rivalapexmediation.adapter.tapjoy",
            "com.rivalapexmediation.adapter.moloco",
            "com.rivalapexmediation.adapter.fyber",
            "com.rivalapexmediation.adapter.smaato",
            "com.rivalapexmediation.adapter.amazon"
        )

        for (packageName in adapterPackages) {
            try {
                val adapterClass = Class.forName("$packageName.Adapter")
                
                // Verify it implements AdNetworkAdapter
                if (AdNetworkAdapter::class.java.isAssignableFrom(adapterClass)) {
                    // Extract network name from package
                    val networkName = packageName.substringAfterLast(".")
                    adapterClasses[networkName] = adapterClass
                }
            } catch (e: ClassNotFoundException) {
                // Adapter not included in build
            }
        }
    }

    private fun isSandboxAdapterDiscoveryEnabled(): Boolean {
        // Enabled in debug builds or when a system property is set (e.g., in CI):
        // -DAPEX_SANDBOX_ADAPTERS=1 → java/system property "apex.sandbox.adapters=1"
        val prop = try {
            System.getProperty("apex.sandbox.adapters")
        } catch (_: Throwable) {
            null
        }
        return BuildConfig.DEBUG || prop == "1"
    }
    
    /**
     * Register custom adapter
     */
    fun registerAdapter(networkName: String, adapterClass: Class<out AdNetworkAdapter>) {
        adapterClasses[networkName] = adapterClass
    }
    
    /**
     * Get adapter instance (lazy initialization)
     */
    fun getAdapter(networkName: String): AdNetworkAdapter? {
        // Check if already initialized
        adapters[networkName]?.let { wrapper ->
            return if (wrapper.isInitialized) wrapper.adapter else null
        }
        
        // Load and initialize adapter
        val adapterClass = adapterClasses[networkName] ?: return null
        
        try {
            val adapter = adapterClass.newInstance() as AdNetworkAdapter
            
            // Check version compatibility
            if (!isCompatible(adapter.minSDKVersion)) {
                adapters[networkName] = AdapterWrapper(
                    adapter = adapter,
                    isInitialized = false,
                    initError = "Incompatible SDK version"
                )
                return null
            }
            
            adapters[networkName] = AdapterWrapper(
                adapter = adapter,
                isInitialized = false
            )
            
            return adapter
        } catch (e: Exception) {
            adapters[networkName] = AdapterWrapper(
                adapter = createDummyAdapter(networkName),
                isInitialized = false,
                initError = e.message
            )
            return null
        }
    }
    
    /**
     * Initialize adapter
     */
    fun initializeAdapter(networkName: String, config: Map<String, Any>) {
        val wrapper = adapters[networkName] ?: return
        
        if (wrapper.isInitialized) {
            return
        }
        
        try {
            wrapper.adapter.initialize(context, config)
            
            adapters[networkName] = wrapper.copy(
                isInitialized = true,
                initError = null
            )
        } catch (e: Exception) {
            adapters[networkName] = wrapper.copy(
                isInitialized = false,
                initError = e.message
            )
        }
    }
    
    /**
     * Check if adapter is initialized
     */
    fun isInitialized(networkName: String): Boolean {
        return adapters[networkName]?.isInitialized == true
    }
    
    /**
     * Get all available adapters
     */
    fun getAvailableAdapters(): List<String> {
        return adapterClasses.keys.toList()
    }
    
    /**
     * Get all initialized adapters
     */
    fun getInitializedAdapters(): List<String> {
        return adapters.filter { it.value.isInitialized }.keys.toList()
    }
    
    /**
     * Diagnostic report of adapter initialization status
     */
    data class InitializationStatus(
        val networkName: String,
        val registered: Boolean,
        val initialized: Boolean,
        val version: String?,
        val minSDKVersion: String?,
        val error: String?
    )

    fun getInitializationReport(): List<InitializationStatus> {
        val known = adapterClasses.keys
        val union = (known + adapters.keys).toSortedSet()
        return union.map { name ->
            val cls = adapterClasses[name]
            val wrapper = adapters[name]
            val adapter: AdNetworkAdapter? = when {
                wrapper != null -> wrapper.adapter
                cls != null -> try { (cls.newInstance() as AdNetworkAdapter) } catch (_: Throwable) { null }
                else -> null
            }
            InitializationStatus(
                networkName = name,
                registered = cls != null,
                initialized = wrapper?.isInitialized == true,
                version = adapter?.version,
                minSDKVersion = adapter?.minSDKVersion,
                error = wrapper?.initError
            )
        }
    }
    
    /**
     * Get adapter initialization error
     */
    fun getAdapterError(networkName: String): String? {
        return adapters[networkName]?.initError
    }
    
    /**
     * Check version compatibility
     */
    private fun isCompatible(minVersion: String): Boolean {
        return compareVersions(sdkVersion, minVersion) >= 0
    }
    
    /**
     * Compare semantic versions
     */
    private fun compareVersions(version1: String, version2: String): Int {
        val parts1 = version1.split(".").map { it.toIntOrNull() ?: 0 }
        val parts2 = version2.split(".").map { it.toIntOrNull() ?: 0 }
        
        for (i in 0 until maxOf(parts1.size, parts2.size)) {
            val part1 = parts1.getOrNull(i) ?: 0
            val part2 = parts2.getOrNull(i) ?: 0
            
            if (part1 != part2) {
                return part1.compareTo(part2)
            }
        }
        
        return 0
    }
    
    /**
     * Create dummy adapter for error handling
     */
    private fun createDummyAdapter(networkName: String): AdNetworkAdapter {
        return object : AdNetworkAdapter {
            override val networkName: String = networkName
            override val version: String = "0.0.0"
            override val minSDKVersion: String = "0.0.0"
            
            override fun initialize(context: Context, config: Map<String, Any>) {}
            
            override fun loadAd(
                placement: String,
                adType: AdType,
                config: Map<String, Any>,
                callback: AdLoadCallback
            ) {
                callback.onAdFailed("Adapter not available", -1)
            }
            
            override fun supportsAdType(adType: AdType): Boolean = false
            
            override fun destroy() {}
        }
    }
    
    /**
     * Clean up all adapters
     */
    fun destroy() {
        adapters.values.forEach { wrapper ->
            try {
                wrapper.adapter.destroy()
            } catch (e: Exception) {
                // Ignore cleanup errors
            }
        }
        
        adapters.clear()
        adapterClasses.clear()
    }
}

/**
 * Example adapter implementation (AdMob)
 */
class AdMobAdapter : AdNetworkAdapter {
    override val networkName = "admob"
    override val version = "1.0.0"
    override val minSDKVersion = "1.0.0"
    
    private var isInitialized = false
    
    override fun initialize(context: Context, config: Map<String, Any>) {
        val appId = config["app_id"] as? String ?: throw IllegalArgumentException("app_id required")
        
        // Initialize AdMob SDK
        // MobileAds.initialize(context, appId)
        
        isInitialized = true
    }
    
    override fun loadAd(
        placement: String,
        adType: AdType,
        config: Map<String, Any>,
        callback: AdLoadCallback
    ) {
        if (!isInitialized) {
            callback.onAdFailed("Adapter not initialized", -1)
            return
        }
        
        // Load ad from AdMob
        // Implementation depends on AdMob SDK integration
    }
    
    override fun supportsAdType(adType: AdType): Boolean {
        return when (adType) {
            AdType.BANNER, AdType.INTERSTITIAL, AdType.REWARDED -> true
            else -> false
        }
    }
    
    override fun destroy() {
        isInitialized = false
    }
}
